### PR TITLE
Redo argument imputation

### PR DESCRIPTION
**Pull request by peterhurford**

### CODE REVIEW

**Comment by peterhurford** [comment]
The test suite grows...

File: R/ensure.R [hunk 1]
```diff
@@ -20,16 +20,21 @@ ensure <- function(fn, preconditions = list(), postconditions = list()) {
...
-      missing_args <- Filter(function(x) { !(x %in% names(args)) }, formals)
+    # Goal here is to (a) impute names the user doesn't give with the formals
+    # and (b) detect if any formals are missing so we can place in their defaults
+    # or error
+    missing_formals <- setdiff(formals, names(args))
+    if (is.null(names(args))) {
+      names(args) <- head(formals, length(args))
+    } else {
+      empty_names <- vapply(names(args), is.empty, logical(1))
+      names(args)[empty_names] <- head(missing_formals, length(args))
```
Comment (robertzk) [inline R/ensure.R:31]: Are you sure this does what you want? It seems like the length of the length-hand side could vary  according as some args are empty but the length of the right-hand side is constant.

**Comment by robertzk** [comment]
Looking good.

File: R/ensure.R (hunk 1 above)
Comment (peterhurford) [inline R/ensure.R:31]: That makes sense, but I can't figure a test that breaks it...

**Pull request merged**